[WIP] Sparkle support

SkyrilHD opened the pull request:
Refer to https://github.com/acidanthera/bugtracker/issues/1627

This pull request adds Sparkle support which still needs to be worked on.

Currently working:
- [x] getting updates
- [x] downloading updates
- [x] added "Check for Updates..."
- [x] added appcast
- [x] added Sparkle to CI
- [x] added automatic updates 

Still needs investigation:
- [x] add signing -> without this, I only had limited testing (installing does not work / error because the signing does not match)
- [x] add appcast automatically (idk if this is possible)
- [x] ~~maybe find a better way for getting Sparkle in CI~~
- [x] ~~only works on RELEASE builds (DEBUG builds will get updated to RELEASE) [so, needs a new way {maybe with CF BundleVersion + (DEBUG/RELEASE) -> `1.6.2 RELEASE`}]~~ bypassed by showing updates only on release builds
- [ ] CFBundleVersion on ${CURRENT_PROJECT_VERSION} causes MaciASL to crash, so I changed it to ${CURRENT_MARKETING_VERSION}
 
There might be questionable design choices such as CI or appcast.xml.

Since Sparkle only works on 10.9 or newer, support for 10.8 has been dropped.

vit9696 commented:
Hello,

Thanks for working on this.

- As for appcast, I could suggest making a separate branch where the CI would commit, once the release is made. Is it feasible? Want to avoid the maintenance need.
- As for DEBUG builds, in my opinion, it is easiest to just disable the updates for them.
- As for translations, do you need them for other languages? Or does Sparkle handle them itself. E.g. in Russian it would be "Проверить обновления…".
- What would you need to test and deploy everything? E.g. from my side.

Vit

vit9696 commented:
Also, I believe api access is prohibited on GitHub CI. You should not use that. Perhaps Sparkle.framework should just be committed to the repository.

SkyrilHD commented:
Hello Vit,

I added the framework to the repository.

- Should be feasible. generate_appcast should do it (it does output an EdDSA signature and packet size), but I am not sure how to implement it. https://sparkle-project.org/documentation/ Check point 5. We need EdDSA signatures, without it, it will fail (at least it did fail on me).
  - should the other branch be empty?
- Is there a way to disable certain MenuBar items? I guess I could disable it that way, if there is no other solution.
- I added the Russian language; afaik Sparkle does not handle them. Speaking of translations, I also need "Check for Updates..." in Chinese as well. I didn't use Google Translate as it can be incorrect.
- The easiest way to test is to downgrade the CFBundleVersion for example to 1.5.8 to simulate an older version of the app. See if it updates the app without an issue

stevezz3 commented:
Hi guys, thank you for working on this!

>  I also need "Check for Updates..." in Chinese as well.

It's `检查更新...`. I am pretty sure it's correct. Love this work and I will keep watching thisˆˆ

SkyrilHD commented:
> > I also need "Check for Updates..." in Chinese as well.
> 
> It's `检查更新...`. I am pretty sure it's correct. Love this work and I will keep watching thisˆˆ

Thank you very much! :)

dhinakg commented:
> Also, I believe api access is prohibited on GitHub CI. You should not use that. Perhaps Sparkle.framework should just be committed to the repository.

API access should work just fine. It is probably a good idea to pin the Sparkle version somehow though.

vit9696 commented:
> It is probably a good idea to pin the Sparkle version somehow though.

Like?

> We need EdDSA signatures, without it, it will fail (at least it did fail on me).

I can make an EdDSA key and provide it to secrets on GitHub. This way the task will be able to sign the dmg images when building an updated appcast.xml. Will that suffice?

> should the other branch be empty?

I guess so.

SkyrilHD commented:
Currently, I am trying to create an appcast, but it fails on me, even though there is a MaciASL DMG:

`Error generating appcast from directory /Users/user/Documents/GitHub/MaciASL/build/Release 
 Error Domain=SUSparkleErrorDomain Code=1001 "No usable archives found in /Users/user/Documents/GitHub/MaciASL/build/Release" UserInfo={NSLocalizedDescription=No usable archives found in /Users/user/Documents/GitHub/MaciASL/build/Release}`

EDIT: Nvm, issue was on my end... I tried to create an appcast from MaciASL without Sparkle implementation.

>I can make an EdDSA key and provide it to secrets on GitHub. This way the task will be able to sign the dmg images when building an updated appcast.xml. Will that suffice?

We definitely need a "private key for signing Sparkle updates". I added mine for testing purposes in secrets. This can be done by pasting the key found in Keychain Access>Login>Private key for signing Sparkle updates. I guess that should be enough. In the comment of "Private key for signing Sparkle updates, you can find a Public key (SUPublicEDKey value). Afaik, that should go to the Info.plist of MaciASL.

dhinakg commented:
> Like?

Specifying it as an env variable on the workflow would do, although the more I think about it just committing Sparkle.framework to the repo is seeming like a better idea as it doesn't break local builds.

SkyrilHD commented:
I think we are almost ready. Generating and pushing appcast works(I think it can be done better). I guess the only thing left is to generate a public and private key with generate_keys: https://github.com/sparkle-project/Sparkle/releases/download/1.26.0/Sparkle-1.26.0.tar.xz (can be found in the bin folder)

The private key goes into secrets called SIGNATURE_FOR_SIGNING (maybe another name suggestion?)
And the public key goes into the info.plist of MaciASL

Generate_appcast has a little limitation (I think) that it can't add Release notes and therefore looks like this:
![image](https://user-images.githubusercontent.com/28839925/133837511-df510f96-a6cf-454f-93fc-ca124de88e2d.png)

Also, is it okay to leave CFBundleVersion as ${CURRENT_MARKETING_VERSION} instead of ${CURRENT_PROJECT_VERSION} (if we leave it as PROJECT_VERSION, the app will crash because apparently Sparkle cannot find the CFBundleVersion. changing to MARKETING_VERSION fixes it)

vit9696 commented on MaciASL/MaciASL-Info-Debug.plist:
```diff
@@ -74,11 +74,11 @@
 	<key>CFBundlePackageType</key>
 	<string>APPL</string>
 	<key>CFBundleShortVersionString</key>
-	<string>${CURRENT_MARKETING_VERSION}</string>
```
That's not right to split the plist files. Especially if one of them stops updating. I believe we should rather have an action to fill `Info.plist` entries if that is particularly needed.

SkyrilHD replied:
Oops, this commit was by accident while I was testing the automatic software update pop up. Should be fixed.

How can we add entries into Info.plist?

vit9696 replied:
One of the ways would be to use `Build Phases`. You can add some shell code there before `Archiving`. For example, you can go with `defaults` or `PlistBuddy`.

SkyrilHD replied:
can we use `plutil` as well?

vit9696 replied:
Yeah

SkyrilHD replied:
Done. Can you check if I did it correctly?

SkyrilHD replied:
"File not found" hmm

SkyrilHD replied:
Fixed "Build" but Clang is complaining... Is there a check if "analyze" is running?

vit9696 commented on .github/workflows/main.yml:
```diff
@@ -70,6 +70,35 @@ jobs:
+      
+      - name: Commit Appcast
+        run: |
+          git config --local user.email "41898282+github-actions[bot]@users.noreply.github.com"
```
What is 41898282? Where does this come from?

SkyrilHD replied:
41898282 can be found in GitHub API: https://api.github.com/users/github-actions%5Bbot%5D

vit9696 replied:
Leave a comment on this, because it feels kind of confusing.

SkyrilHD replied:
I guess this is not necessary. However, if I recall correctly, the person how commits would be who did the release. I could remove it though

vit9696 replied:
Well, it makes sense to show that the appcast is committed by a bot. Just I want an explanation on what the bot is like "The 41898282 identifier comes from the GitHub Actions API: https://api.github.com/users/github-actions%5Bbot%5D."

SkyrilHD replied:
tbh, I just looked what the e-mail address for the GitHub actions bot is and pasted it there.

vit9696 commented on .github/workflows/main.yml:
```diff
@@ -70,6 +70,35 @@ jobs:
+      - uses: actions/checkout@v2
+      
+      - name: Get Tag
+        run: echo "TAG_VER=${GITHUB_REF/refs\/tags\//}" >> $GITHUB_ENV
```
Is there any particular reason to expose `TAG_VER` through GitHub environment? I only see it being used in shell scripts below. Sounds like it can be just a normal variable.

SkyrilHD replied:
Fixed.

vit9696 commented on .github/workflows/main.yml:
```diff
@@ -70,6 +70,35 @@ jobs:
+      
+      - name: Get Information & Sign
+        run: |
+           echo "DATE=$(date -R)" >> $GITHUB_ENV
```
Same for DATE and SIGNATURE really. Why cannot it be e.g.

```sh
TAG_VER=${GITHUB_REF/refs\/tags\//}
DATE=$(date -R)
SIGNATURE=$(./bin/sign_update -s ${{ secrets.SIGNATURE_FOR_SIGNING }} build/Release/MaciASL-${TAG_VER}-RELEASE.dmg)
./bin/generate_appcast -o appcast.xml -s ${{ secrets.SIGNATURE_FOR_SIGNING }} --download-url-prefix https://github.com/acidanthera/MaciASL/releases/download/${TAG_VER}/ build/Release
```

or something similar? Also, this should do `|| exit 1` on failure, no?

SkyrilHD replied:
Fixed and added

vit9696 commented on MaciASL/MaciASL-Info-Release.plist:
```diff
@@ -0,0 +1,186 @@
+	<true/>
+	<key>SUFeedURL</key>
+	<string>https://raw.githubusercontent.com/acidanthera/MaciASL/appcast/appcast.xml</string>
+	<key>#SUPublicEDKey</key>
```
I need to make a key pair and post the public one here, right?

SkyrilHD replied:
Yes, exactly.

vit9696 replied:
```
    <key>SUPublicEDKey</key>
    <string>d6TTl9OeRP+CH+IiQwgLCUsxj6DXK+Fnrcfmcj61ANk=</string>
```

SkyrilHD replied:
added

vit9696 commented on MaciASL/iASL.m:
```diff
@@ -235,6 +235,11 @@ +(void)applicationWillFinishLaunching:(NSNotification *)notification {
         IOObjectRelease(expert);
         NSMenu *acpi = [[[NSApp mainMenu] itemAtIndex:1] submenu];
         [[acpi insertItemWithTitle:NSLocalizedString(@"new-from-acpi", @"New from ACPI") action:NULL keyEquivalent:@"" atIndex:1] setSubmenu:menu];
+#ifndef DEBUG
```
Why not just `#ifdef DEBUG`?

SkyrilHD replied:
Fixed.

vit9696 reviewed (commented):
Where do `bin` files come from? Like `generate_appcast` and `sign_update`?

SkyrilHD commented:
> Where do bin files come from? Like generate_appcast and sign_update?

The files are from Sparkle 1.26: https://github.com/sparkle-project/Sparkle/releases/download/1.26.0/Sparkle-1.26.0.tar.xz

vit9696 commented:
> The files are from Sparkle 1.26: https://github.com/sparkle-project/Sparkle/releases/download/1.26.0/Sparkle-1.26.0.tar.xz

I think we should download them with CI instead. It is not too great to distribute binary files that are only optionally needed and only for the CI itself, not for normal users.

vit9696 commented on .github/workflows/main.yml:
```diff
@@ -70,6 +70,35 @@ jobs:
+      - uses: actions/checkout@v2
+      
+      - name: Get Tag
+        run: TAG_VER=${GITHUB_REF/refs\/tags\//}
```
That will not work actually. Needs to be done inside `Get Information & Sign`.

SkyrilHD replied:
Noticed and fixed

vit9696 commented on .github/workflows/main.yml:
```diff
@@ -70,6 +70,35 @@ jobs:
+        run: |
+           DATE=$(date -R)
+           SIGNATURE=$(./bin/sign_update -s ${{ secrets.SIGNATURE_FOR_SIGNING }} build/Release/MaciASL-${{ env.TAG_VER }}-RELEASE.dmg)
+           ./bin/generate_appcast -o appcast.xml -s ${{ secrets.SIGNATURE_FOR_SIGNING }} --download-url-prefix https://github.com/acidanthera/MaciASL/releases/download/${{ env.TAG_VER }}/ build/Release || exit 1
```
`${{ env.TAG_VER }}` will not be available. It should be `${TAG_VER}` here and above.

SkyrilHD replied:
fixed

vit9696 commented on .github/workflows/main.yml:
```diff
@@ -70,6 +70,35 @@ jobs:
+          git checkout --orphan appcast
+          git add appcast.xml
+          git commit appcast.xml -m "Update Appcast"
+          git push --set-upstream origin appcast --force
```
Does it need any special permissions to do pushes?

SkyrilHD replied:
as in?

vit9696 replied:
GitHub configuration or something like that. I do not know how it works at the moment.

SkyrilHD replied:
It did work for me without modifying. Currently looks like this: https://github.com/SkyrilHD/MaciASL/tree/appcast

vit9696 commented on .github/workflows/main.yml:
```diff
@@ -70,6 +70,33 @@ jobs:
+           TAG_VER=${GITHUB_REF/refs\/tags\//}
+           DATE=$(date -R)
+           SIGNATURE=$(./bin/sign_update -s ${{ secrets.SIGNATURE_FOR_SIGNING }} build/Release/MaciASL-${TAG_VER}-RELEASE.dmg)
+           ./bin/generate_appcast -o appcast.xml -s ${{ secrets.SIGNATURE_FOR_SIGNING }} --download-url-prefix https://github.com/acidanthera/MaciASL/releases/download/${TAG_VER}/ build/Release || exit 1
```
Is there no mistake here? I do not see `SIGNATURE` being used anywhere. If it is not needed and creates some file, it is best to have `./bin/sign_update -s ${{ secrets.SIGNATURE_FOR_SIGNING }} build/Release/MaciASL-${TAG_VER}-RELEASE.dmg`.

SkyrilHD replied:
Sorry, overlooked

vit9696 replied:
It probably needs some error checking too.

SkyrilHD replied:
`- name: Get Sparkle 1.26
        run: |
          curl -L -s "https://github.com/sparkle-project/Sparkle/releases/download/1.26.0/Sparkle-1.26.0.tar.xz" -o Sparkle.tar.xz
          tar -xf Sparkle.tar.xz`

This one too?

vit9696 replied:
Yes. `|| exit 1` feels good enough.

vit9696 commented on .github/workflows/main.yml:
```diff
@@ -70,6 +70,38 @@ jobs:
+      - name: Get Sparkle 1.26
+        run: |
+          curl -L -s "https://github.com/sparkle-project/Sparkle/releases/download/1.26.0/Sparkle-1.26.0.tar.xz" -o Sparkle.tar.xz || exit 1
+          tar -xf Sparkle.tar.xz
```
Needs an `|| exit 1`?

SkyrilHD replied:
I thought it probably is not needed as extracting only works if the file is downloaded successfully. But thinking about it, an error check is probably a good idea.

vit9696 commented on MaciASL.xcodeproj/xcshareddata/xcschemes/MaciASL.xcscheme:
```diff
@@ -0,0 +1,88 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
This seems to be committed by accident.

SkyrilHD replied:
Yep, sorry for that.

vit9696 commented on MaciASL.xcodeproj/project.pbxproj:
```diff
@@ -458,6 +478,24 @@
+			);
+			runOnlyForDeploymentPostprocessing = 0;
+			shellPath = /bin/sh;
+			shellScript = "if [ \"$CONFIGURATION\" == \"Release\" ]; then\nplutil -insert \"SUAllowsAutomaticUpdates\" -bool true build/Release/MaciASL.app/Contents/Info.plist\nplutil -insert \"SUEnableAutomaticChecks\" -bool true build/Release/MaciASL.app/Contents/Info.plist\nplutil -insert \"SUFeedURL\" -string \"https://raw.githubusercontent.com/acidanthera/MaciASL/appcast/appcast.xml\" build/Release/M ...
```
- Is it `Release` or `RELEASE`?
- Can running build multiple times result in duplicates?
- Do not use relative paths like `build/Release`, there is `${TARGET_BUILD_DIR}` variable.

SkyrilHD replied:
> Is it Release or RELEASE?

In archiving script I saw it being `Release`, so I used that as well.

> Can running build multiple times result in duplicates?

No, I just tried it locally and it does output this error for example: `Build/Products/Release/MaciASL.app/Contents/Info.plist: Could not modify plist, error: Value true already exists at key path SUAllowsAutomaticUpdates`


Xcode outputs this error `Command PhaseScriptExecution failed with a nonzero exit code`

> Do not use relative paths like build/Release, there is ${TARGET_BUILD_DIR} variable.

Thanks! I am still learning.

SkyrilHD commented:
@vit9696 I see the appcast ci gave out an error. Probably due to this: `Deleting the contents of '/Users/runner/work/MaciASL/MaciASL'`

In this case, I think we need to remove this line: https://github.com/acidanthera/MaciASL/blob/11a2a7bc71e736d92db775f6ce6d5252199cefba/.github/workflows/main.yml#L83

what do you think?

vit9696 commented:
Thanks for a hint. Tried, waiting for to build. To be honest, I get a feeling it must be one job, not two to work.

SkyrilHD commented:
Yep, you are right. It starts another runner in which the dmg is not included

vit9696 commented:
Could you submit an update on that later please? Thanks!
